File: AIReaderQt/src/Match/MatchView.cpp
#include "Match/MatchView.hpp"

#include "Common/Tappable.hpp"
#include "Common/Ui.hpp"

#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QVBoxLayout>

namespace {

Tappable* card(const QString& html, std::function<void()> pick) {
    auto* card = new Tappable(std::move(pick));
    card->setProperty("card", true);
    auto* layout = new QVBoxLayout(card);
    layout->setContentsMargins(12, 10, 12, 10);
    layout->addWidget(Ui::rich(html));
    return card;
}

std::string status(const MatchRound& round) {
    std::string misses = round.misses() == 0 ? "no misses"
        : std::to_string(round.misses()) + (round.misses() == 1 ? " miss" : " misses");
    if (round.isComplete()) return "All " + std::to_string(round.matchedCount()) + " paired, " + misses + ".";
    std::string progress = std::to_string(round.matchedCount()) + " of " + std::to_string(round.cards().size()) + " paired, " + misses + ".";
    if (!round.lastPick()) return round.matchedCount() == 0 ? "Click a word, then the meaning it had." : progress;
    return (round.lastPick()->matched ? "✓ A pair. " : "✕ Not a pair. ") + progress;
}

}  // namespace

MatchView::MatchView(Env& env, Navigator& navigator, long long bookId)
    : Screen(navigator, "Practice"), feature_(env, bookId) {
    auto* holder = new QWidget;
    QVBoxLayout* column = Ui::column(holder);
    status_ = new QLabel;
    columns_ = new QHBoxLayout;
    columns_->setSpacing(10);
    next_ = new QPushButton("Next round");
    QObject::connect(next_, &QPushButton::clicked, this, [this] { feature_.nextRound(); });
    column->addWidget(status_);
    column->addLayout(columns_);
    column->addWidget(next_, 0, Qt::AlignLeft);
    setBody(Ui::scrolled(holder));

    feature_.onChange = [this] { render(); };
    render();
}

void MatchView::render() {
    if (round_ != feature_.roundNumber()) build();
    refresh();
}

void MatchView::build() {
    Ui::clear(columns_);
    fronts_.clear();
    backs_.clear();
    round_ = feature_.roundNumber();
    if (!feature_.hasRound()) {
        columns_->addWidget(Ui::label("Look up two words or more while reading, then come back to pair them with their meanings."));
        return;
    }
    const MatchRound& round = feature_.round();
    auto* fronts = new QVBoxLayout;
    auto* backs = new QVBoxLayout;
    for (int index : round.fronts()) {
        const Card& front = round.cards()[index];
        QString html = "<b>" + Ui::escape(front.front) + "</b>";
        if (!front.lemma.empty()) html += "<br>" + Ui::small(Ui::escape(front.lemma));
        fronts_.push_back(card(html, [this, index] { feature_.pickFront(index); }));
        fronts->addWidget(fronts_.back());
    }
    for (int index : round.backs()) {
        const Card& back = round.cards()[index];
        QString html = Ui::escape(back.back);
        if (!back.example.empty()) html += "<br>" + Ui::small("<i>" + Ui::escape(back.example) + "</i>");
        backs_.push_back(card(html, [this, index] { feature_.pickBack(index); }));
        backs->addWidget(backs_.back());
    }
    columns_->addLayout(fronts, 2);
    columns_->addLayout(backs, 3);
}

void MatchView::refresh() {
    status_->setVisible(feature_.hasRound());
    next_->setVisible(feature_.hasRound() && feature_.round().isComplete());
    if (!feature_.hasRound()) return;
    const MatchRound& round = feature_.round();
    for (size_t i = 0; i < fronts_.size(); ++i) {
        int index = round.fronts()[i];
        fronts_[i]->setSelected(round.selectedFront() == index);
        fronts_[i]->setEnabled(!round.isMatched(index));
    }
    for (size_t i = 0; i < backs_.size(); ++i) {
        int index = round.backs()[i];
        backs_[i]->setSelected(round.selectedBack() == index);
        backs_[i]->setEnabled(!round.isMatched(index));
    }
    status_->setText(Ui::q(status(round)));
}

File: AIReader/AIReader/Features/Match/MatchView.swift
import ComposableArchitecture
import SwiftUI

/// The board: words down one column, meanings down the other. A tap fills a
/// card until its pair is chosen; a matched pair fades out.
struct MatchView: View {
    let store: StoreOf<MatchFeature>

    var body: some View {
        Group {
            if let round = store.round {
                board(round)
            } else if store.isLoaded {
                ContentUnavailableView(
                    "Nothing to practise yet",
                    systemImage: "rectangle.on.rectangle.angled",
                    description: Text("Look up two words or more while reading, then come back to pair them with their meanings.")
                )
            } else {
                ProgressView()
            }
        }
        .navigationTitle("Practice")
        .task { await store.send(.task).finish() }
    }

    private func board(_ round: MatchRound) -> some View {
        ScrollView {
            VStack(spacing: 16) {
                Text(status(round))
                    .font(.footnote)
                    .foregroundStyle(.secondary)
                    .frame(maxWidth: .infinity, alignment: .leading)

                HStack(alignment: .top, spacing: 10) {
                    VStack(spacing: 10) {
                        ForEach(round.fronts, id: \.self) { card in
                            tile(selected: round.selectedFront == card, matched: round.isMatched(card)) {
                                store.send(.frontTapped(card))
                            } label: {
                                front(round.cards[card])
                            }
                        }
                    }
                    .frame(width: 130)

                    VStack(spacing: 10) {
                        ForEach(round.backs, id: \.self) { card in
                            tile(selected: round.selectedBack == card, matched: round.isMatched(card)) {
                                store.send(.backTapped(card))
                            } label: {
                                back(round.cards[card])
                            }
                        }
                    }
                }

                if round.isComplete {
                    Button("Next round") { store.send(.nextRoundTapped) }
                        .buttonStyle(.borderedProminent)
                }
            }
            .padding()
        }
    }

    private func tile(
        selected: Bool,
        matched: Bool,
        action: @escaping () -> Void,
        @ViewBuilder label: () -> some View
    ) -> some View {
        Button(action: action) {
            label()
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(10)
                .background(selected ? Color.accentColor.opacity(0.2) : Color.secondary.opacity(0.15))
                .clipShape(RoundedRectangle(cornerRadius: 10))
                .overlay {
                    RoundedRectangle(cornerRadius: 10)
                        .strokeBorder(selected ? Color.accentColor : .clear, lineWidth: 2)
                }
        }
        .buttonStyle(.plain)
        .disabled(matched)
        .opacity(matched ? 0.25 : 1)
        .animation(.default, value: matched)
    }

    private func front(_ card: Card) -> some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(card.front).font(.headline)
            if let lemma = card.lemma {
                Text(lemma).font(.caption).foregroundStyle(.secondary)
            }
        }
    }

    private func back(_ card: Card) -> some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(card.back).font(.subheadline)
            if !card.example.isEmpty {
                Text(card.example).font(.caption).italic().foregroundStyle(.secondary)
            }
        }
    }

    private func status(_ round: MatchRound) -> String {
        let misses = round.misses == 0 ? "no misses" : "\(round.misses) miss\(round.misses == 1 ? "" : "es")"
        if round.isComplete { return "All \(round.matchedCount) paired, \(misses)." }
        let progress = "\(round.matchedCount) of \(round.cards.count) paired, \(misses)."
        guard let pick = round.lastPick else {
            return round.matchedCount == 0 ? "Tap a word, then the meaning it had." : progress
        }
        return (pick.matched ? "✓ A pair. " : "✕ Not a pair. ") + progress
    }
}
